From 53f474e6a8d74d5dc0c3a015d889471f9a157685 Mon Sep 17 00:00:00 2001
From: Johannes Berg <johannes.berg@intel.com>
Date: Fri, 1 Jun 2018 09:45:55 +0200
Subject: [PATCH] iwlwifi: pcie: gen2: build A-MSDU only for GSO
Git-commit: 53f474e6a8d74d5dc0c3a015d889471f9a157685
Patch-mainline: v4.20-rc1
References: bsc#1051510

If the incoming frame should be an A-MSDU, it may already be one,
for example in the case of NAN multicast being encapsulated in an
A-MSDU. Thus, use the GSO algorithm to build A-MSDU only if the
skb actually contains GSO data.

Fixes: 6ffe5de35b05 ("iwlwifi: pcie: add AMSDU to gen2")
Signed-off-by: Johannes Berg <johannes.berg@intel.com>
Signed-off-by: Luca Coelho <luciano.coelho@intel.com>
Acked-by: Takashi Iwai <tiwai@suse.de>

---
 drivers/net/wireless/intel/iwlwifi/pcie/tx-gen2.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/intel/iwlwifi/pcie/tx-gen2.c
+++ b/drivers/net/wireless/intel/iwlwifi/pcie/tx-gen2.c
@@ -414,7 +414,12 @@ struct iwl_tfh_tfd *iwl_pcie_gen2_build_
 
 	hdr_len = ieee80211_hdrlen(hdr->frame_control);
 
-	if (amsdu) {
+	/*
+	 * Only build A-MSDUs here if doing so by GSO, otherwise it may be
+	 * an A-MSDU for other reasons, e.g. NAN or an A-MSDU having been
+	 * built in the higher layers already.
+	 */
+	if (amsdu && skb_shinfo(skb)->gso_size) {
 		if (iwl_pcie_gen2_build_amsdu(trans, skb, tfd,
 					      tb1_len + IWL_FIRST_TB_SIZE,
 					      hdr_len, dev_cmd))
